### PR TITLE
Azure fix log collector

### PR DESCRIPTION
**Pull request by soyacz**
collect-logs does not work for unified provisioning and logs are not correctly gathered from Azure nodes.
This PR fixes that.

Besides this fix, some refactoring for better unit-testing SCT:
1. We can easily create fake resources (in memory) allowing testing of methods that operate on VM's
2. provisioner tests are applicable to any cloud provider (need only minimal adaptation in fixture preparation)

Also unified InstancesDefinition building - making easy to reuse by all cloud providers (yet maintaining flexibility).

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Review by fgelcer — APPROVED** [review]
LGTM

File: sdcm/logcollector.py [hunk 1]
```diff
@@ -1233,6 +1235,29 @@ def get_gce_instances_by_testid(self):
...
+                                                   "key_file": f"~/.ssh/{instance.ssh_key_name}"},
+                                               instance=instance,
+                                               global_ip=instance.public_ip_address,
+                                               tags=instance.tags) for instance in instances]
+            self.db_cluster = [c_node for c_node in collecting_nodes if c_node.tags.get(
+                "NodeType") in ("scylla-db", "oracle-db")]
+            self.monitor_set = [c_node for c_node in collecting_nodes if c_node.tags.get("NodeType") == "monitor"]
+            self.loader_set = [c_node for c_node in collecting_nodes if c_node.tags.get("NodeType") == "loader"]
+            self.kubernetes_set = [c_node for c_node in collecting_nodes
+                                   if c_node.tags.get("NodeType") == "k8s"]  # why aws and gce uses 'loader'? Bug?
```
Comment (fruch) [inline sdcm/logcollector.py:1255]: I don't think we have a this `NodeType==k8s` defined anywhere
@vponomaryov correct me if i'm wrong.

File: sdcm/logcollector.py [hunk 2]
```diff
@@ -1233,6 +1235,29 @@ def get_gce_instances_by_testid(self):
         if self.params["use_cloud_manager"]:
             self.find_and_append_cloud_manager_instance_to_collecting_nodes()
 
+    def create_collecting_nodes(self):
+        try:
+            provisioners = provisioner_factory.discover_provisioners(backend=self.backend, test_id=self.test_id)
+            instances = sum([provisioner.list_instances() for provisioner in provisioners], [])
+            collecting_nodes = [CollectingNode(name=instance.name,
+                                               ssh_login_info={
+                                                   "hostname": instance.public_ip_address,
```
Comment (fruch) [inline sdcm/logcollector.py:1244]: this function is in theory, would be the same one for aws/gce (once their provision would match), right ?

is so, please use public only if it's says so on the `param.get('ip_ssh_connections')`

**Review by fruch — APPROVED** [review]
LGTM

File: sdcm/logcollector.py (hunk 2 above)
Comment (fruch) [reply sdcm/logcollector.py:1244]: But we'll fix it once we'll get to it

**Comment by fruch** [comment]
@soyacz 
seem like those changes broke the azure artifact test:
https://jenkins.scylladb.com/job/scylla-master/job/artifacts/job/artifacts-azure-image-test/28/

```
11:58:28      self._definition_builder = region_definition_builder.get_builder(self.params, test_config=self.test_config)
11:58:28  AttributeError: 'ScyllaAzureCluster' object has no attribute 'params'
```

**Pull request merged**